source/common/router/router.cc

```diff
@@ -746,6 +746,14 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::RequestHeaderMap& headers,
       !transport_socket_options_ || !transport_socket_options_->http11ProxyInfo().has_value();
   UpstreamRequestPtr upstream_request = std::make_unique<UpstreamRequest>(
       *this, std::move(generic_conn_pool), can_send_early_data, can_use_http3);
+  if (retry_state_ && retry_state_->shouldRetryOnConnectionFailure() &&
+      Runtime::runtimeFeatureEnabled("envoy.restart_features.ensure_connection_retry")) {
+    upstream_request->addUpstreamCallbacks(*this);
+    wait_for_connect_ = true;
+    retry_connection_failure_buffer_limit_ =
+        std::max(retry_shadow_buffer_limit_, 2 * cluster_->perConnectionBufferLimitBytes());
+    callbacks_->setDecoderBufferLimit(retry_connection_failure_buffer_limit_);
```

Comment:
we are now permanently increasing this buffer limit in this case, seems like we might want lower it back to its original value when we establish the connection. Alternatively, we could have a "connect buffer" where this is buffered in the case of connect and don't otherwise use this buffer

Reply:
I would prefer not using a 3rd buffer, this would most likely have a none negligible impact on the memory and CPU performance due to the copy of the body until we get a connect.

I am fine with lowering the buffer limit to its original value once we get a connect.
I did not do it yet due to the recommendation in https://github.com/envoyproxy/envoy/blob/1b768b49f9a85a865c35735a4944a001975055e9/envoy/http/filter.h#L699-L701.

I guess this should be fine as the router filter should be the last stream filter in the filter chain. And that from a search in the code base, all filter call `setDecoderBufferLimit` in the `decodeHeaders` method.
I added a comment in https://github.com/envoyproxy/envoy/blob/main/envoy/http/filter.h.

Reply:
There are custom filters that you might not see e.g. private that can call this during decodeBody, etc. 

Using a 3rd buffer isn't actually that bad because the underlying buffer object would just move around the slices (e.g. the backing store) so moving from the filter chain buffer to a connect buffer isn't that bad.

cc @alyssawilk  for additional thoughts

Reply:
From my understanding the retry and shadow limit will most likely be equal to the FM buffer limit:
https://github.com/DataDog/envoy/blob/dfbcf58749b47bc8f98241d78023975da0b94f34/source/common/router/router.cc#L1000
```
retry_shadow_buffer_limit_ = callbacks_->decoderBufferLimit();
```

I do not think we can go above it see: https://github.com/envoyproxy/envoy/pull/32359#discussion_r1526455610
so we would need to increase it or did I missing something ?